pkg/apis/apps/kfdef/v1beta1/application_types.go

```diff
@@ -0,0 +1,211 @@
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package v1beta1
+
+import (
+	"k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+)
+
+const (
+	KfConfigFile = "app.yaml"
+)
+
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// KfDef is the Schema for the applications API
+// +k8s:openapi-gen=true
+type KfDef struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	Spec   KfDefSpec   `json:"spec,omitempty"`
+	Status KfDefStatus `json:"status,omitempty"`
+}
+
+type KfDefSpec struct {
+	Applications []Application `json:"applications,omitempty"`
+	Plugins      []Plugin      `json:"plugins,omitempty"`
+	Secrets      []Secret      `json:"secrets,omitempty"`
+	Repos        []Repo        `json:"repos,omitempty"`
+}
+
+// Application defines an application to install
+type Application struct {
+	Name            string           `json:"name,omitempty"`
+	KustomizeConfig *KustomizeConfig `json:"kustomizeConfig,omitempty"`
+}
+
+type KustomizeConfig struct {
+	RepoRef    *RepoRef    `json:"repoRef,omitempty"`
+	Overlays   []string    `json:"overlays,omitempty"`
+	Parameters []NameValue `json:"parameters,omitempty"`
+}
+
+type RepoRef struct {
+	Name string `json:"name,omitempty"`
+	Path string `json:"path,omitempty"`
+}
+
+// ParamType indicates the type of an input parameter;
+// Used to distinguish between a single string and an array of strings.
+type ParamType string
+
+// Valid ParamTypes:
+const (
+	ParamTypeString ParamType = "string"
+	ParamTypeArray  ParamType = "array"
+)
+
+// ArrayOrString is a type that can hold a single string or string array.
+// Used in JSON unmarshalling so that a single JSON field can accept
+// either an individual string or an array of strings.
+type ArrayOrString struct {
+	Type      ParamType // Represents the stored type of ArrayOrString.
+	StringVal string
+	ArrayVal  []string
+}
+
+// UnmarshalJSON implements the json.Unmarshaller interface.
+func (arrayOrString *ArrayOrString) UnmarshalJSON(value []byte) error {
+	if value[0] == '"' {
+		arrayOrString.Type = ParamTypeString
+		return json.Unmarshal(value, &arrayOrString.StringVal)
+	}
+	arrayOrString.Type = ParamTypeArray
+	return json.Unmarshal(value, &arrayOrString.ArrayVal)
+}
+
+// MarshalJSON implements the json.Marshaller interface.
+func (arrayOrString ArrayOrString) MarshalJSON() ([]byte, error) {
+	switch arrayOrString.Type {
+	case ParamTypeString:
+		return json.Marshal(arrayOrString.StringVal)
+	case ParamTypeArray:
+		return json.Marshal(arrayOrString.ArrayVal)
+	default:
+		return []byte{}, fmt.Errorf("impossible ArrayOrString.Type: %q", arrayOrString.Type)
+	}
+}
+
+type NameValue struct {
+	Name  string        `json:"name,omitempty"`
+	Value ArrayOrString `json:"value,omitempty"`
```

Comment:
Why did this field become array or string?
Do we have a use case for this?

Reply:
@kkasravi

Reply:
tektoncd has a similar construct

within the config file you may want to pass data that represents a dictionary
eg
      parameters:
      - name: cpu-devices
        value:
        - skylake
        - broadwell
        - crest
      - name: tpu-devices
        - nvidia-tesla-t4
        - nvidia-tesla-p100
        - nvidia-tesla-t4-vws

Reply:
ArrayOrString seems unnecessarily complicated

```
type ArrayOrString struct {
	Type      ParamType // Represents the stored type of ArrayOrString.
	StringVal string
	ArrayVal  []string
}
```

If we want to support arrays why not just do

```
type NameValue struct {
     Name string
     Value[] string
}
```

A single string argument could be handled by the case where array is size one.

But given how we are using parameters does it even make sense to support arrays? We are using kustomize parameters and creating key-value pairs. Doesn't that only handle string scalars? What would it mean for kustomize to substitute an Array into the manifests.

My vote would be to optimize for simplicity. I think our existing use cases are covered by scalar strings

```
type NameValue struct {
     Name string
     Value string
}
```

If we don't have a use case for Arrays then I suggest we stick with that. If we need to support more types in the future we can do that e.g.

```
type NameValue struct {
     Name string
     Value string
     ArrayValue []string
     MapValue map[string]string
}
```

Reply:
From @kkasravi 
> If we allow lists to be passed to kf apps then params.env is not the best way to pass a list.
This approach (string or array-of-string) is used by tektoncd in their parameter passing implementation. We should try and have well typed information for app params

How would a list let help users specify environment variables?